No SecRandomCopyBytes before 10.7, so existing code is broken by design.
Below we borrow from OpenBSD implementation which uses getentropy, which is available
via legacy-support. This is not tested, as of now.

--- a/vendor/rand_os/src/macos.rs	2019-11-05 01:34:46.000000000 +0800
+++ b/vendor/rand_os/src/macos.rs	2025-11-12 09:07:55.000000000 +0800
@@ -14,40 +14,27 @@
 use super::OsRngImpl;
 
 use std::io;
-use self::libc::{c_int, size_t};
 
 #[derive(Clone, Debug)]
 pub struct OsRng;
 
-enum SecRandom {}
-
-#[allow(non_upper_case_globals)]
-const kSecRandomDefault: *const SecRandom = 0 as *const SecRandom;
-
-#[link(name = "Security", kind = "framework")]
-extern {
-    fn SecRandomCopyBytes(rnd: *const SecRandom,
-                          count: size_t, bytes: *mut u8) -> c_int;
-}
-
 impl OsRngImpl for OsRng {
     fn new() -> Result<OsRng, Error> { Ok(OsRng) }
 
     fn fill_chunk(&mut self, dest: &mut [u8]) -> Result<(), Error> {
         let ret = unsafe {
-            SecRandomCopyBytes(kSecRandomDefault,
-                               dest.len() as size_t,
-                               dest.as_mut_ptr())
+            libc::getentropy(dest.as_mut_ptr() as *mut libc::c_void, dest.len())
         };
         if ret == -1 {
-            Err(Error::with_cause(
+            return Err(Error::with_cause(
                 ErrorKind::Unavailable,
-                "couldn't generate random bytes",
-                io::Error::last_os_error()))
-        } else {
-            Ok(())
+                "getentropy failed",
+                io::Error::last_os_error()));
         }
+        Ok(())
     }
 
-    fn method_str(&self) -> &'static str { "SecRandomCopyBytes" }
+    fn max_chunk_size(&self) -> usize { 256 }
+
+    fn method_str(&self) -> &'static str { "getentropy" }
 }
